apple-dmg/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+        for chunk in bytes.chunks(2048 * 512) {
+            let mut encoder = ZlibEncoder::new(chunk, Compression::best());
+            let mut compressed = vec![];
+            encoder.read_to_end(&mut compressed)?;
+            let compressed_length = compressed.len() as u64;
+            let sector_count = chunk.len() as u64 / 512;
+            self.w.write_all(&compressed)?;
+            self.data_hasher.update(&compressed);
+            table.add_chunk(BlkxChunk::new(
+                ChunkType::Zlib,
+                self.sector_number,
+                sector_count,
+                self.compressed_offset,
+                compressed_length,
+            ));
+            self.sector_number += sector_count;
+            self.compressed_offset += compressed_length;
+        }
+        table.add_chunk(BlkxChunk::term(self.sector_number, self.compressed_offset));
+        self.main_hasher.update(&table.checksum.data[..4]);
+        self.xml
+            .add_partition(Partition::new(id as i32 - 1, name, table));
+        Ok(())
+    }
+
+    pub fn finish(mut self) -> Result<()> {
+        let mut xml = vec![];
+        plist::to_writer_xml(&mut xml, &self.xml)?;
+        let pos = self.w.seek(SeekFrom::Current(0))?;
+        let data_digest = self.data_hasher.finalize();
+        let main_digest = self.main_hasher.finalize();
+        let koly = KolyTrailer::new(
+            pos,
+            self.sector_number,
+            pos,
+            xml.len() as _,
+            data_digest,
+            main_digest,
+        );
+        self.w.write_all(&xml)?;
+        koly.write_to(&mut self.w)?;
+        Ok(())
+    }
+}
+
+fn symlink(target: &str) -> Vec<u8> {
+    let xsym = format!(
+        "XSym\n{:04}\n{:x}\n{}\n",
+        target.as_bytes().len(),
+        md5::compute(target.as_bytes()),
+        target,
+    );
+    let mut xsym = xsym.into_bytes();
+    xsym.resize(1067, b' ');
+    xsym
+}
+
+fn add_dir<T: ReadWriteSeek>(src: &Path, dest: &Dir<'_, T>) -> Result<()> {
+    for entry in std::fs::read_dir(src)? {
+        let entry = entry?;
+        let file_name = entry.file_name();
+        let file_name = file_name.to_str().unwrap();
+        let source = src.join(&file_name);
+        let file_type = entry.file_type()?;
+        if file_type.is_dir() {
+            let d = dest.create_dir(file_name)?;
+            add_dir(&source, &d)?;
+        } else if file_type.is_file() {
+            let mut f = dest.create_file(file_name)?;
+            std::io::copy(&mut File::open(source)?, &mut f)?;
+        } else if file_type.is_symlink() {
+            let target = std::fs::read_link(&source)?;
+            let xsym = symlink(target.to_str().unwrap());
+            let mut f = dest.create_file(file_name)?;
+            std::io::copy(&mut &xsym[..], &mut f)?;
+        }
+    }
+    Ok(())
+}
+
+pub fn create_dmg(dir: &Path, dmg: &Path, volume_label: &str, total_sectors: u32) -> Result<()> {
```

Comment:
We'll probably need to refactor this into a type to facilitate customizing all the options. But this is fine as a first implementation!